src/main/java/org/openrewrite/analysis/dataflow/ExternalFlowModels.java

```diff
@@ -159,56 +181,111 @@ private AdditionalFlowStepPredicate forFlowFromArgumentIndexToQualifier(
                             callMatcher.advanced().isParameter(srcCursor, argumentIndex);
         }
 
-        private List<AdditionalFlowStepPredicate> optimize(Collection<FlowModel> models) {
-            Map<Integer, List<FlowModel>> flowFromArgumentIndexToReturn = new HashMap<>();
-            Map<Integer, List<FlowModel>> flowFromArgumentIndexToQualifier = new HashMap<>();
+        private AdditionalFlowStepPredicate forFlowFromArgumentIndexToArgumentIndex(
+                ArgumentIndices argumentIndices,
+                Collection<? extends InvocationMatcher> methodMatchers
+        ) {
+            InvocationMatcher callMatcher = InvocationMatcher.from(methodMatchers);
+            if (argumentIndices.inputIndex == -1) {
+                return (srcExpression, srcCursor, sinkExpression, sinkCursor) ->
+                        callMatcher.advanced().isSelect(srcCursor) &&
+                                callMatcher.advanced().isParameter(sinkCursor, argumentIndices.outputIndex);
+            } else {
+                return (srcExpression, srcCursor, sinkExpression, sinkCursor) ->
+                        callMatcher.advanced().isParameter(srcCursor, argumentIndices.inputIndex) &&
+                                callMatcher.advanced().isParameter(sinkCursor, argumentIndices.outputIndex);
+            }
+        }
+
+        @Value
+        static class PredicateToFlowModels {
+            AdditionalFlowStepPredicate predicate;
+            Set<FlowModel> models;
+        }
+
+        @Data
+        static class ArgumentIndices {
+            final int inputIndex;
+            final int outputIndex;
```

Comment:
Do these need to be `final` or are they automatically final because of the `@Data` annotation?